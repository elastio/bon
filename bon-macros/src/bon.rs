use crate::builder;
use crate::normalization::{ExpandCfg, ExpansionOutput};
use crate::util::prelude::*;

pub(crate) fn generate(params: TokenStream2, item: TokenStream2) -> TokenStream2 {
    try_generate(params, item.clone())
        .unwrap_or_else(|err| crate::error::error_into_token_stream(err, item))
}

pub(crate) fn try_generate(params: TokenStream2, item: TokenStream2) -> Result<TokenStream2> {
    let item: syn::Item = syn::parse2(item)?;
    let macro_path = syn::parse_quote!(::bon::bon);

    let ctx = ExpandCfg {
        macro_path,
        params,
        item,
    };

    let (_params, item) = match ctx.expand_cfg()? {
        ExpansionOutput::Expanded { params, item } => (params, item),
        ExpansionOutput::Recurse(output) => return Ok(output),
    };

    match item {
        syn::Item::Impl(item_impl) => builder::item_impl::generate(item_impl),
        _ => bail!(
            &item,
            "`#[bon]` attribute is expected to be placed on an `impl` block \
             but it was placed on other syntax instead"
        ),
    }
}
