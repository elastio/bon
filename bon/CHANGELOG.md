# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/elastio/bon/compare/bon-v0.1.0...bon-v0.1.1) - 2024-07-23

### Other
- Upd docs
- More improvements, deploy-website workflow
- Docs for alternatives plus `#[builder(name = ...)]` for members
- Rename `field -> member`
- Add more docs
- Add doctests, more docs
- More docs
- More docs and better decisions
- Fix builder attrs not being removed correctly
- Next iteration
- Test struct input
- Start testing struct input
- Refactor in preparation for struct input for macro
- Make `#[default]` compatible with `Option`
- Add `default` and `into` attrs
- Make bool optional, strip `_` prefix
- Add special casing for `Option` setters
- Better doc comments
- Improve nested refs handling
- Cleanup
- Fix a bunch of papercuts with codegen
- Finish implementing support for associated functions
- Start adding support for associated functions
- Add some docs
- Rebranding buildy -> prox (buildy crate name is already taken on crates.io)
