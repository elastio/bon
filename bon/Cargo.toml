[package]
name    = "bon"
version = "2.3.0"

description = "Next-gen compile-time-checked builder generator, named function's arguments, and more!"

categories = [
  "rust-patterns",
  "data-structures",
  "asynchronous",
  "no-std",
  "no-std::no-alloc",
]
keywords = ["builder", "macro", "derive", "constructor", "setter"]

edition    = "2021"
homepage   = "https://bon-rs.com/"
license    = "MIT OR Apache-2.0"
repository = "https://github.com/elastio/bon"

# This MSRV was chosen because Rust supports mixing generic type and const
# parameters only starting with this version. We require this feature for the
# cases when the builder is generated for a function or struct that uses const
# generics. The generated builder **always** appends a generic type parameter
# for the type state at the end of the Builder generics list. So for functions
# or structs that use const generics the generated builder will have a generic
# type parameter after the const generic parameter, which is only supported
# starting from Rust 1.59.0.
rust-version = "1.59.0"

[package.metadata.docs.rs]
all-features = true

# Generate clickable links in the source code view in the docs
rustdoc-args = ["--generate-link-to-definition"]

# We don't need the docs to be built for every first-tier target.
# This crate is not platform-specific.
targets = ["x86_64-unknown-linux-gnu"]

[lints]
workspace = true

[dependencies]
# The version of the macro crate is pinned to a specific one because the code
# generated by the macros uses private APIs from the runtime crate that are not
# guarded by semver.
bon-macros  = { path = "../bon-macros", version = "=2.3.0" }
rustversion = "1"

[dev-dependencies]
# Using a bit older version that supports our MSRV
expect-test = "1.4.1"

# Using a bit older version that supports our MSRV
tokio = { version = "1.29.1", features = ["macros", "rt-multi-thread"] }

# Using a bit older version that supports our MSRV
trybuild = "1.0.89"

[features]
alloc   = []
default = ["std"]
std     = ["alloc"]

# Opts in to the higher MSRV 1.79.0. In this version, Rust stabilized the syntax
# for bounds in associated type position which can be used to make bounds on generic
# associated types implied. See the release announcement for more:
# https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.htmlbounds-in-associated-type-position
#
# This feature is useful for the trait `IsComplete` generated by the builder macros.
# When this feature is enabled, the builder macros use the new syntax for bounds in
# associated type position, which enables implied `IsSet` bounds for the type state
# of required members.
#
# To understand how this can be used consider the following example:
#
# ```rust
# #[derive(bon::Builder)]
# struct Example {
#     a: u32,
#     b: Option<u32>,
# }
#
# use example_builder::{IsUnset, IsComplete};
#
# impl<State: example_builder::State> ExampleBuilder<State> {
#     fn build_with_default_b(self) -> Example
#     where
#         State: IsComplete,
#         State::B: IsUnset,
#     {
#         self.b(42).build()
#     }
# }
# ```
#
# This code wouldn't compile without this feature enabled, because `State: IsComplete`
# wouldn't automatically imply `State::A: IsSet`, so the builder type state returned
# after `self.b()` doesn't imply that the member `a` is set, and thus `build()`
# can't be called.
#
# Huge thanks to @harudagondi for suggesting the name of this cargo feature!
implied-bounds = ["bon-macros/implied-bounds"]

# üî¨ Experimental! There may be breaking changes to this feature between *minor* releases,
# however, compatibility within patch releases is guaranteed though.
#
# This feature enables the #[builder(overwritable)] attribute that can be used to
# allow overwriting already set fields in the builder. Without this attribute it's
# not possible to set the same member twice in the builder, a compile error would be
# generated.
#
# See more info at https://bon-rs.com/reference/builder/top-level/overwritable.
#
# We are considering stabilizing this attribute if you have a use for it. Please leave
# a üëç reaction under the issue https://github.com/elastio/bon/issues/149 if you need
# this attribute. It would also be cool if you could leave a comment under that issue
# describing your use case for it.
experimental-overwritable = ["bon-macros/experimental-overwritable"]
