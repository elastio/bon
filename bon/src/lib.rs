#![doc = include_str!("../README.md")]

pub use bon_macros::*;

/// Symbols used by macros. They are not stable and are considered an implementation detail.
#[doc(hidden)]
pub mod private;

/// Same as [`std::vec!`] but converts each element with [`Into::into()`].
///
/// **WARNING:** it's not recommended to import this macro into scope. Reference it
/// using the full path (`bon::vec![]`) to avoid confusion with the [`std::vec!`] macro.
///
/// A good example of the use case for this macro is when you want to create a
/// [`Vec<String>`] where part of the items are hardcoded string literals of type
/// `&str` and the other part is made of dynamic [`String`] values.
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: Vec<String> = bon::vec![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `vec![expr; N]` syntax, since it's simpler to
/// just write `vec![expr.into(); N]` using [`std::vec!`] instead.
#[macro_export]
macro_rules! vec {
    () => (::std::vec::Vec::new());
    ($($item:expr),+ $(,)?) => (::std::vec![$(::core::convert::Into::into($item)),+ ]);
}

/// Creates a fixed-size array literal where each element is converted with [`Into::into()`]
/// into the target type.
///
/// You'll probably need a hint for the target type of items in the array if the
/// compiler can't infer it from its usage.
///
/// This is similar in spirit to the [`bon::vec!`] macro, but it's for arrays.
/// See [`bon::vec!`] docs for details.
///
/// Same example as in [`bon::vec!`], but using this macro. It works with array
/// as well because [`Command::args`] accepts any value that implements [`IntoIterator`]:
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: [String; 4] = bon::arr![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `[expr; N]` syntax, since it's simpler to
/// just write `[expr.into(); N]` instead.
///
/// [`Command::args`]: std::process::Command::args
/// [`bon::vec!`]: crate::vec
#[macro_export]
macro_rules! arr {
    () => ([]);
    ($($item:expr),+ $(,)?) => ([$(::core::convert::Into::into($item)),+]);
}

#[cfg(test)]
mod tests {
    #[test]
    fn arr_smoke() {
        let actual: [String; 3] = crate::arr!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: [String; 0] = crate::arr![];
        assert!(actual.is_empty());
    }

    #[test]
    fn vec_smoke() {
        let actual: Vec<String> = crate::vec!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: Vec<String> = crate::vec![];
        assert!(actual.is_empty());
    }
}
