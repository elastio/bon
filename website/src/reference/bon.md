# `#[bon]`

This is a companion macro for [`builder`]. You should place it on top of the `impl` block
where you want to define methods with the [`builder`] macro. It provides
the necessary context to the [`builder`] macros on top of the functions
inside of the `impl` block. You'll get compile errors without that context.

For the examples of the usage of this macro and the reason why it's needed, see this paragraph in [the overview](../guide/overview#builder-for-an-associated-method).

[`builder`]: ./builder
