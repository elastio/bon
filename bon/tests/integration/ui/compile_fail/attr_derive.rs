use bon::{bon, builder, Builder};

struct NoTraitImpls;

#[derive(Builder)]
#[builder(derive(Clone, Debug))]
struct StructContainsNonTrait {
    #[builder(start_fn)]
    no_impl_start_fn: NoTraitImpls,

    no_impls_required: NoTraitImpls,

    no_impl_optional: Option<NoTraitImpls>,

    #[builder(default = NoTraitImpls)]
    no_impl_optional_2: NoTraitImpls,

    x: u32,
}

#[builder(derive(Clone, Debug))]
fn fn_contains_non_trait(
    #[builder(start_fn)] //
    _no_impl_start_fn: NoTraitImpls,

    _no_impls_required: NoTraitImpls,

    _no_impl_optional: Option<NoTraitImpls>,

    #[builder(default = NoTraitImpls)] //
    _no_impl_optional_2: NoTraitImpls,

    _x: u32,
) {
}

#[bon]
impl StructContainsNonTrait {
    #[builder(derive(Clone, Debug))]
    fn method_contains_non_trait(
        self,

        #[builder(start_fn)] _no_impl_start_fn: NoTraitImpls,

        _no_impls_required: NoTraitImpls,

        _no_impl_optional: Option<NoTraitImpls>,

        #[builder(default = NoTraitImpls)] //
        _no_impl_optional_2: NoTraitImpls,

        _x: u32,
    ) {
    }
}

#[derive(Builder)]
#[builder(derive())]
struct EmptyDerives {}

#[derive(Builder)]
#[builder(derive(Clone()))]
struct EmptyParamsForDerive {}

#[derive(Builder)]
#[builder(derive(Clone(bounds {})))]
struct WrongDelimInBounds {}

#[builder(derive(Into))]
fn derive_into_with_finish_fn_member(#[builder(finish_fn)] _finish_fn: fn() -> u32) -> u32 {
    99
}

#[builder(derive(Into))]
fn function_with_unit_return_type() {}

#[builder(derive(Into))]
async fn async_function() -> u32 {
    99
}

#[builder(derive(Into))]
unsafe fn unsafe_function() -> u32 {
    99
}

#[builder(derive(Into))]
async unsafe fn unsafe_async_function() -> u32 {
    99
}

struct MethodWithUnsupportedIntoDerive;

#[bon]
impl MethodWithUnsupportedIntoDerive {
    #[builder(derive(Into))]
    fn derive_into_with_finish_fn_member(#[builder(finish_fn)] _finish_fn: fn() -> u32) -> u32 {
        99
    }
}

#[bon]
impl MethodWithUnsupportedIntoDerive {
    #[builder(derive(Into))]
    fn function_with_unit_return_type() {}
}

#[bon]
impl MethodWithUnsupportedIntoDerive {
    #[builder(derive(Into))]
    async fn async_function() -> Self {
        Self
    }
}

#[bon]
impl MethodWithUnsupportedIntoDerive {
    #[builder(derive(Into))]
    unsafe fn unsafe_function() -> Self {
        Self
    }
}

#[bon]
impl MethodWithUnsupportedIntoDerive {
    #[builder(derive(Into))]
    async unsafe fn unsafe_async_function() -> Self {
        Self
    }
}

#[builder(derive(Into(bounds(u32: Copy))))]
fn unsupported_bounds_on_derive_into() -> u32 {
    99
}

fn main() {}
