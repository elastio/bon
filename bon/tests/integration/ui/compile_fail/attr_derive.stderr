error: expected at least one parameter in parentheses
  --> tests/integration/ui/compile_fail/attr_derive.rs:59:11
   |
59 | #[builder(derive())]
   |           ^^^^^^

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
 --> tests/integration/ui/compile_fail/attr_derive.rs:9:23
  |
9 |     no_impl_start_fn: NoTraitImpls,
  |                       ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
  |
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
  |
3 + #[derive(Clone)]
4 | struct NoTraitImpls;
  |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:11:24
   |
11 |     no_impls_required: NoTraitImpls,
   |                        ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:13:30
   |
13 |     no_impl_optional: Option<NoTraitImpls>,
   |                              ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:16:25
   |
16 |     no_impl_optional_2: NoTraitImpls,
   |                         ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
 --> tests/integration/ui/compile_fail/attr_derive.rs:9:23
  |
9 |     no_impl_start_fn: NoTraitImpls,
  |                       ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
  |
  = help: the trait `Debug` is not implemented for `NoTraitImpls`
  = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
 --> src/private/derives.rs
  |
  | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
  |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
  |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:11:24
   |
11 |     no_impls_required: NoTraitImpls,
   |                        ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:13:30
   |
13 |     no_impl_optional: Option<NoTraitImpls>,
   |                              ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:16:25
   |
16 |     no_impl_optional_2: NoTraitImpls,
   |                         ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:24:24
   |
24 |     _no_impl_start_fn: NoTraitImpls,
   |                        ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:26:25
   |
26 |     _no_impls_required: NoTraitImpls,
   |                         ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:28:31
   |
28 |     _no_impl_optional: Option<NoTraitImpls>,
   |                               ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:31:26
   |
31 |     _no_impl_optional_2: NoTraitImpls,
   |                          ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:24:24
   |
24 |     _no_impl_start_fn: NoTraitImpls,
   |                        ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:26:25
   |
26 |     _no_impls_required: NoTraitImpls,
   |                         ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:28:31
   |
28 |     _no_impl_optional: Option<NoTraitImpls>,
   |                               ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:31:26
   |
31 |     _no_impl_optional_2: NoTraitImpls,
   |                          ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `StructContainsNonTrait: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:38:6
   |
38 | impl StructContainsNonTrait {
   |      ^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `StructContainsNonTrait`

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:44:28
   |
44 |         _no_impl_start_fn: NoTraitImpls,
   |                            ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:46:29
   |
46 |         _no_impls_required: NoTraitImpls,
   |                             ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:48:35
   |
48 |         _no_impl_optional: Option<NoTraitImpls>,
   |                                   ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: the trait bound `NoTraitImpls: Clone` is not satisfied
  --> tests/integration/ui/compile_fail/attr_derive.rs:51:30
   |
51 |         _no_impl_optional_2: NoTraitImpls,
   |                              ^^^^^^^^^^^^ the trait `Clone` is not implemented for `NoTraitImpls`
   |
note: required by a bound in `clone_member`
  --> src/private/derives.rs
   |
   | pub fn clone_member<T: Clone>(member: &Option<T>) -> Option<T> {
   |                        ^^^^^ required by this bound in `clone_member`
help: consider annotating `NoTraitImpls` with `#[derive(Clone)]`
   |
3  + #[derive(Clone)]
4  | struct NoTraitImpls;
   |

error[E0277]: `StructContainsNonTrait` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:38:6
   |
38 | impl StructContainsNonTrait {
   |      ^^^^^^^^^^^^^^^^^^^^^^ `StructContainsNonTrait` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `StructContainsNonTrait`
   = note: add `#[derive(Debug)]` to `StructContainsNonTrait` or manually `impl Debug for StructContainsNonTrait`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:44:28
   |
44 |         _no_impl_start_fn: NoTraitImpls,
   |                            ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:46:29
   |
46 |         _no_impls_required: NoTraitImpls,
   |                             ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:48:35
   |
48 |         _no_impl_optional: Option<NoTraitImpls>,
   |                                   ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |

error[E0277]: `NoTraitImpls` doesn't implement `Debug`
  --> tests/integration/ui/compile_fail/attr_derive.rs:51:30
   |
51 |         _no_impl_optional_2: NoTraitImpls,
   |                              ^^^^^^^^^^^^ `NoTraitImpls` cannot be formatted using `{:?}`
   |
   = help: the trait `Debug` is not implemented for `NoTraitImpls`
   = note: add `#[derive(Debug)]` to `NoTraitImpls` or manually `impl Debug for NoTraitImpls`
note: required by a bound in `as_dyn_debug`
  --> src/private/derives.rs
   |
   | pub fn as_dyn_debug<T: Debug>(member: &T) -> &dyn Debug {
   |                        ^^^^^ required by this bound in `as_dyn_debug`
help: consider annotating `NoTraitImpls` with `#[derive(Debug)]`
   |
3  + #[derive(Debug)]
4  | struct NoTraitImpls;
   |
